File: shared/otel/go/otelkit/temporal.go
package otelkit

import (
	"go.opentelemetry.io/otel"
	temporalotel "go.temporal.io/sdk/contrib/opentelemetry"
	"go.temporal.io/sdk/interceptor"
)

// TemporalInterceptors returns the Temporal worker tracing interceptor from
// go.temporal.io/sdk/contrib/opentelemetry (SPEC §2.5). Wire it via
// worker.Options{Interceptors: []interceptor.WorkerInterceptor{...}}.
//
// Note: the SPEC names the return type interceptor.WorkerInterceptors
// (plural); no such type exists in go.temporal.io/sdk v1.16–v1.33 — the
// worker-consumable interface is interceptor.WorkerInterceptor, which the
// tracing interceptor satisfies (interceptor.Interceptor embeds it).
//
// The interceptor uses the ambient tracer provider and W3C propagator, so it
// is automatically a no-op when OTEL_SDK_DISABLED left the default no-op
// providers in place. If interceptor construction fails (invalid options),
// nil is returned and the worker runs untraced.
func TemporalInterceptors() interceptor.WorkerInterceptor {
	ti, err := temporalotel.NewTracingInterceptor(temporalotel.TracerOptions{
		Tracer:            otel.Tracer(tracerName + "/temporal"),
		TextMapPropagator: otel.GetTextMapPropagator(),
	})
	if err != nil {
		return nil
	}
	return ti
}

File: shared/otel/python/otelkit/temporal.py
"""Temporal workflow tracing interceptor (SPEC §2.5, §3 Temporal row).

Uses temporalio's contrib OpenTelemetry interceptor
(``temporalio.contrib.opentelemetry.TracingInterceptor``) which propagates
W3C tracecontext through workflow/activity headers.
"""

from __future__ import annotations

import logging
from typing import Any, Optional

logger = logging.getLogger("otelkit.temporal")


def temporal_interceptor(tracer: Optional[Any] = None) -> Optional[Any]:
    """Return the temporalio OpenTelemetry TracingInterceptor, or None.

    Returns None (logged, never raises) when temporalio or its opentelemetry
    contrib extra is not installed, or when OTEL_SDK_DISABLED is set.
    Pass the result to ``temporalio.client.Client(..., interceptors=[...])``
    and ``temporalio.worker.Worker(..., interceptors=[...])``.
    """
    from . import sdk_disabled

    if sdk_disabled():
        return None
    try:
        from temporalio.contrib.opentelemetry import TracingInterceptor

        if tracer is None:
            from . import get_tracer

            tracer = get_tracer("otelkit.temporal")
        return TracingInterceptor(tracer)
    except Exception as exc:
        logger.warning("otelkit: temporal interceptor unavailable (%s)", exc)
        return None
